src/develop/pixelpipe_hb.c

```diff
@@ -2547,6 +2564,142 @@ float *dt_dev_get_raster_mask(const dt_dev_pixelpipe_t *pipe, const dt_iop_modul
   return raster_mask;
 }
 
+void dt_dev_clear_luminance_mask(dt_dev_pixelpipe_t *pipe)
+{
+  if(pipe->luminance_mask_data) dt_free_align(pipe->luminance_mask_data);
+  pipe->luminance_mask_data = NULL;
+}
+
+gboolean dt_dev_write_luminance_mask(dt_dev_pixelpipe_iop_t *piece, float *const rgb, const dt_iop_roi_t *const roi_in, const int mode)
+{
+  dt_dev_pixelpipe_t *p = piece->pipe;
+  if((p->want_luminance_mask & DT_DEV_LUMINANCE_MASK_REQUIRED) == 0) return FALSE;
+  if((p->want_luminance_mask & ~DT_DEV_LUMINANCE_MASK_REQUIRED) != mode) return FALSE;
+
+  dt_dev_clear_luminance_mask(p);
+
+  const int width = roi_in->width;
+  const int height = roi_in->height;
+  float *mask = dt_alloc_align_float((size_t)width * height);
+  if(mask == NULL) return TRUE;
+
+  p->luminance_mask_data = mask;
+  memcpy(&p->luminance_mask_roi, roi_in, sizeof(dt_iop_roi_t));
+
+  dt_masks_calc_luminance_mask(rgb, mask, width, height);
+  return FALSE;
+}
+
+#ifdef HAVE_OPENCL
+gboolean dt_dev_write_luminance_mask_cl(dt_dev_pixelpipe_iop_t *piece, cl_mem in, const dt_iop_roi_t *const roi_in, const int mode)
+{
+  dt_dev_pixelpipe_t *p = piece->pipe;  
+
+  if((p->want_luminance_mask & DT_DEV_LUMINANCE_MASK_REQUIRED) == 0) return FALSE;
+  if((p->want_luminance_mask & ~DT_DEV_LUMINANCE_MASK_REQUIRED) != mode) return FALSE;
+
+  dt_dev_clear_luminance_mask(p);
+
+  const int width = roi_in->width;
+  const int height = roi_in->height;
+
+  cl_mem out = NULL;
+  float *mask = NULL;
+  const int devid = p->devid;
+
+  mask = dt_alloc_align_float((size_t)width * height);
+  if(mask == NULL) goto error;
+  out = dt_opencl_alloc_device(devid, width, height, sizeof(float));   
+  if(out == NULL) goto error;
+
+  const int program = 31;
+  int kernel_calc_luminance_mask = dt_opencl_create_kernel(program, "out_luminance_mask");
```

Comment:
we usually create a kernel once and use it later. Maybe this should be cached somewhere. maybe moving it into `dt_develop_blend_init_cl_global()`.

Reply:
right.